compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -67,6 +67,37 @@ where
            impl2_def_id={:?})",
         impl1_def_id, impl2_def_id,
     );
+    // Before doing expensive operations like entering an inference context, do
+    // a quick check via fast_reject to tell if the impl headers could possibly
+    // unify.
+    let impl1_self = tcx.type_of(impl1_def_id);
+    let impl2_self = tcx.type_of(impl2_def_id);
+    let impl1_ref = tcx.impl_trait_ref(impl1_def_id);
+    let impl2_ref = tcx.impl_trait_ref(impl2_def_id);
+
+    // Check if any of the input types definitely do not unify.
+    if impl1_ref
+        .iter()
+        .flat_map(|tref| tref.substs.types())
+        .zip(impl2_ref.iter().flat_map(|tref| tref.substs.types()))
+        .chain(iter::once((impl1_self, impl2_self)))
```

Comment:
that isn't required, is it?

The impl_self should be `tref.substs[0]`, so you're currently checking that one twice